packages/ui/src/elements/RenderServerComponent/WithHOC.tsx

```diff
@@ -0,0 +1,38 @@
+import { isReactServerComponentOrFunction } from 'payload/shared'
+
+import { RenderClientHOC } from './RenderClientHOC.js'
+
+/**
+ * Client components cannot be invoked on the server and must be done on the client. Here are some rules:
+ * - If the HOC a server component, it can be invoked on the server (HOC()). Its component to render can be either a server or client component.
+ * - If the HOC is a client component, it must be rendered a component on the client (<HOC />). Its component to render MUST be a client component OR a pre-rendered server component.
+ */
+export const WithHOC: React.FC<{
+  Component: React.ComponentType
+  componentKey?: string
+  HOC?: (Component: React.ComponentType, RenderedComponent?: React.FC) => React.ComponentType
+  isRSC: boolean
+  props: object
+}> = ({ Component, componentKey, HOC, isRSC, props }) => {
+  let ComponentToRender = Component
+
+  const HOCisRSC = isReactServerComponentOrFunction(HOC)
+
+  if (HOCisRSC) {
+    ComponentToRender = HOC(Component)
+  } else if (!HOCisRSC) {
+    if (isRSC) {
+      return (
+        <RenderClientHOC
```

Comment:
I think this could be a function call too? Not sure here

Reply:
This particular instance is not able to be invoked from the server because it's a client component. It must be rendered instead. Here's the exact error:

```plaintext
Attempted to call RenderClientHOC() from the server but RenderClientHOC is on the client. It's not possible to invoke a client function from the server, it can only be rendered as a Component or passed to props of a Client Component."
```